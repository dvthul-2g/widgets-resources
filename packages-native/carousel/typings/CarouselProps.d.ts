/**
 * This file was generated from Carousel.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix Widgets Team
 */
import { ListValue, ObjectItem } from "mendix";
import { ReactNode } from "react";

interface CommonProps<Style> {
    name: string;
    style: Style[];
}

export type LayoutEnum = "card" | "fullWidth";

export type ActiveSlideAlignmentEnum = "center" | "start";

export interface CarouselProps<Style> extends CommonProps<Style> {
    contentSource: ListValue;
    content: (item: ObjectItem) => ReactNode;
    layout: LayoutEnum;
    showPagination: boolean;
    activeSlideAlignment: ActiveSlideAlignmentEnum;
    loop: boolean;
}
