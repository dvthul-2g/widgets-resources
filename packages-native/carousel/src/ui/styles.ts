import { TextStyle, ViewStyle } from "react-native";

interface DotStyle {
    container?: ViewStyle;
    opacity?: number;
    scale?: number;
    color?: string;
}

export interface ActiveDotStyle extends ViewStyle, DotStyle {}

interface ContainerStyle extends ViewStyle {
    width?: number;
    height?: number;
}

interface LayoutStyle {
    container?: ViewStyle;
    carousel?: ContainerStyle;
    slideItem?: {
        opacity?: number;
        scale?: number;
    };
    activeSlideItem?: ViewStyle;
    paginationContainer?: ViewStyle;
    paginationText?: TextStyle;
    dotStyle?: ActiveDotStyle;
    activeDotStyle?: ActiveDotStyle;
}

export interface CarouselStyle {
    cardLayout: LayoutStyle;
    fullWidthLayout: LayoutStyle;
}

const sharedStyles: LayoutStyle = {
    container: {
        width: "100%"
    },
    paginationContainer: {
        paddingHorizontal: 0,
        paddingVertical: 0
    },
    activeDotStyle: {
        marginHorizontal: 0
    },
    dotStyle: {
        container: {
            marginHorizontal: 0
        },
        marginHorizontal: 0
    }
};

export const defaultCarouselStyle: CarouselStyle = {
    cardLayout: {
        ...sharedStyles,
        activeSlideItem: {
            width: "70%"
        }
    },
    fullWidthLayout: {
        ...sharedStyles,
        activeSlideItem: {
            width: "100%"
        },
        slideItem: {
            opacity: 1,
            scale: 1
        }
    }
};
