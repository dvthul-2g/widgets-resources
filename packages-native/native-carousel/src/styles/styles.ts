import { TextStyle, ViewStyle } from "react-native";

interface DotStyle {
    container?: ViewStyle;
    opacity?: number;
    scale?: number;
    color?: string;
}

export interface ActiveDotStyle extends ViewStyle, DotStyle {}

interface ContainerStyle extends ViewStyle {
    width?: number;
    height?: number;
}

interface LayoutStyle {
    container?: ViewStyle;
    carousel?: ContainerStyle;
    slideItem?: {
        opacity?: number;
        scale?: number;
    };
    activeSlideItem?: ViewStyle;
    paginationContainer?: ViewStyle;
    paginationText?: TextStyle;
    dotStyle?: ActiveDotStyle;
    activeDotStyle?: ActiveDotStyle;
}

export interface CarouselStyle {
    cardLayout: LayoutStyle;
    fullWidthLayout: LayoutStyle;
}

export const defaultCarouselStyle: CarouselStyle = {
    cardLayout: {
        container: {
            width: "100%"
        },
        activeSlideItem: {
            width: "70%"
        },
        paginationContainer: {
            marginHorizontal: 0
        },
        activeDotStyle: {
            marginHorizontal: 0
        },
        dotStyle: {
            container: {
                marginHorizontal: 0
            },
            marginHorizontal: 0
        }
    },
    fullWidthLayout: {
        container: {
            width: "100%"
        },
        activeSlideItem: {
            width: "100%"
        },
        slideItem: {
            opacity: 1,
            scale: 1
        },
        paginationContainer: {
            marginHorizontal: 0
        },
        activeDotStyle: {
            marginHorizontal: 0
        },
        dotStyle: {
            container: {
                marginHorizontal: 0
            },
            marginHorizontal: 0
        }
    }
};
