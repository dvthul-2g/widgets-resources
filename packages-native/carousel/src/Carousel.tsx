import { createElement, Fragment, useCallback, useEffect, useRef, useState } from "react";
import { ActivityIndicator, LayoutChangeEvent, Text, View } from "react-native";
import { CarouselProps } from "../typings/CarouselProps";
import { CarouselStyle, defaultCarouselStyle, LayoutStyle } from "./ui/styles";
import { default as NativeCarousel, Pagination } from "react-native-snap-carousel";
import deepmerge from "deepmerge";
import { ObjectItem, ValueStatus } from "mendix";

export const Carousel = (props: CarouselProps<CarouselStyle>): JSX.Element => {
    const carousel = useRef<any>(null);
    const [sliderDimensions, setSliderDimensions] = useState({
        slider: { width: 0, height: 0 },
        slide: { width: 0, height: 0 }
    });

    const customStyles = props.style ? props.style.filter(o => o != null) : [];

    const styles = deepmerge.all<CarouselStyle>([defaultCarouselStyle, ...customStyles]);

    const layoutSpecificStyle: LayoutStyle =
        props.layout === "fullWidth" ? styles.fullWidthLayout! : styles.cardLayout!;

    const [activeSlide, setActiveSlide] = useState(0);

    const [loading, setLoading] = useState(true);

    useEffect(() => {
        if (
            props.contentSource?.status === ValueStatus.Available &&
            props.contentSource.items?.length &&
            props.contentSource.items.length > 0
        ) {
            setLoading(false);
        }
    }, [props.contentSource]);

    const onSnap = useCallback((index: number) => {
        setActiveSlide(index);
    }, []);

    const renderItem = ({ item, index }: { item: ObjectItem; index: number }) => {
        return (
            <View key={index} style={{ ...layoutSpecificStyle.slideItem }}>
                {props.content(item)}
            </View>
        );
    };

    const renderPagination = useCallback(() => {
        if (!props.showPagination) {
            return null;
        }

        const contentLength = props.contentSource.items!.length;
        const paginationOverflow = contentLength > 5;
        const { pagination } = layoutSpecificStyle;

        if (paginationOverflow) {
            return (
                <View style={pagination.container}>
                    <Text style={pagination.text}>
                        {activeSlide + 1}/{contentLength}
                    </Text>
                </View>
            );
        }

        return (
            <Pagination
                dotsLength={contentLength}
                activeDotIndex={activeSlide}
                containerStyle={pagination.container}
                dotContainerStyle={pagination.dotContainerStyle}
                dotColor={pagination.dotStyle?.color}
                dotStyle={pagination.dotStyle}
                inactiveDotStyle={pagination.inactiveDotStyle}
                inactiveDotColor={pagination.inactiveDotStyle?.color}
                inactiveDotOpacity={pagination.inactiveDotStyle?.opacity}
                inactiveDotScale={pagination.inactiveDotStyle?.scale}
                carouselRef={carousel.current}
                tappableDots
            />
        );
    }, [activeSlide, carousel.current, props.contentSource, props.showPagination]);

    const onLayout = (event: LayoutChangeEvent) => {
        let viewHeight = event.nativeEvent.layout.height;
        const viewWidth = event.nativeEvent.layout.width;

        let itemWidth = 0;
        let itemHeight = 0;

        if (layoutSpecificStyle.slideItem) {
            const { width: slideItemWidth, height: slideItemHeight } = layoutSpecificStyle.slideItem;
            // We calculate the actual number value in order to
            // allow users to set width and height as percentage since lib only accepts numbers

            if (typeof slideItemWidth === "string" && slideItemWidth.includes("%")) {
                const percentage = +slideItemWidth.replace("%", "");
                itemWidth = (viewWidth * percentage) / 100;
            } else {
                itemWidth = Number(slideItemWidth);
            }
            if (typeof slideItemHeight === "string" && slideItemHeight.includes("%")) {
                const percentage = +slideItemHeight.replace("%", "");
                itemHeight = (viewWidth * percentage) / 100;
            } else {
                itemHeight = Number(slideItemHeight);
            }

            // We don't want to pass the already processed height to the item container
            delete layoutSpecificStyle.slideItem.height;

            if (styles.container?.height === undefined && itemHeight > 0) {
                viewHeight = itemHeight;
            }
        }

        setSliderDimensions({
            slider: { width: viewWidth, height: viewHeight },
            slide: { width: itemWidth, height: itemHeight }
        });
    };

    return (
        <View style={styles.container} onLayout={onLayout}>
            {loading ? (
                <ActivityIndicator color={layoutSpecificStyle.indicator!.color} size="large" />
            ) : (
                /* library ignores width/height if its vertical/horizontal */
                sliderDimensions.slide.width > 0 &&
                sliderDimensions.slider.width > 0 && (
                    <Fragment>
                        <NativeCarousel
                            loop={props.loop}
                            useScrollView
                            activeSlideAlignment={props.activeSlideAlignment}
                            layout="default"
                            firstItem={0}
                            initialNumToRender={props.contentSource.items?.length}
                            data={props.contentSource.items ?? []}
                            renderItem={renderItem}
                            sliderWidth={sliderDimensions.slider.width}
                            sliderHeight={sliderDimensions.slider.height}
                            itemWidth={sliderDimensions.slide.width}
                            itemHeight={sliderDimensions.slide.height}
                            inactiveSlideScale={layoutSpecificStyle.inactiveSlideItem?.scale}
                            inactiveSlideOpacity={layoutSpecificStyle.inactiveSlideItem?.opacity}
                            onSnapToItem={onSnap}
                            ref={carousel}
                        />
                        {renderPagination()}
                    </Fragment>
                )
            )}
        </View>
    );
};
