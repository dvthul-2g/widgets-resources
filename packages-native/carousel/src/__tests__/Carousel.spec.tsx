import { CarouselProps } from "../../typings/CarouselProps";
import { CarouselStyle } from "../ui/styles";
import { render } from "react-native-testing-library";
import { createElement } from "react";
import { Text, View } from "react-native";
import { ListValueBuilder } from "@widgets-resources/piw-utils";
import { Carousel } from "../Carousel";

jest.mock("react-native", () => ({
    InteractionManager: require.requireActual("InteractionManager"),
    Platform: require.requireActual("Platform"),
    Animated: require.requireActual("Animated"),
    I18nManager: require.requireActual("I18nManager"),
    View: require.requireActual("View"),
    Text: require.requireActual("Text"),
    ViewPropTypes: {
        style: jest.fn()
    },
    TouchableOpacity: require.requireActual("TouchableOpacity"),
    Easing: require.requireActual("Easing"),
    FlatList: require.requireActual("FlatList"),
    ScrollView: require.requireActual("ScrollView"),
    StyleSheet: require.requireActual("StyleSheet"),
    Image: require.requireActual("Image"),
    ActivityIndicator: require.requireActual("ActivityIndicator")
}));

jest.mock("prop-types", () => require.requireActual("prop-types"));

describe("Carousel", () => {
    let defaultProps: CarouselProps<CarouselStyle>;
    const listValueBuilder = ListValueBuilder();
    beforeEach(() => {
        defaultProps = {
            name: "carousel",
            contentSource: listValueBuilder.simple(),
            content: jest.fn(() => (
                <View>
                    <Text>MyContent</Text>
                </View>
            )),
            layout: "card",
            showPagination: true,
            activeSlideAlignment: "center",
            loop: false,
            style: []
        };
    });

    /* TODO: enzyme or react-native-testing-lib still not allowing mocking states with hooks
        Thus we can't snapshot test the actual carousel
     */

    it("renders loading", () => {
        expect(render(<Carousel {...defaultProps} />).toJSON()).toMatchSnapshot();
    });

    // it("renders without pagination", () => {
    //     expect(render(<Carousel {...defaultProps} showPagination={false} />).toJSON()).toMatchSnapshot();
    // });
    //
    // it("renders full width", () => {
    //     expect(render(<Carousel {...defaultProps} layout={"fullWidth"} />).toJSON()).toMatchSnapshot();
    // });
    //
    // it("renders numbered pagination if item size is more than 5", () => {
    //     const props = {
    //         ...defaultProps,
    //         contentSource: listValueBuilder.withAmountOfItems(6)
    //     };
    //     expect(render(<Carousel {...props} />).toJSON()).toMatchSnapshot();
    // });
});
