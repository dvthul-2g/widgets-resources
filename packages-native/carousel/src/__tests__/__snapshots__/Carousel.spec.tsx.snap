// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel renders loading 1`] = `
<RCTView
  onLayout={[Function]}
  style={
    Object {
      "width": "100%",
    }
  }
/>
`;
